code/modules/food_and_drink/cakes.dm

```diff
@@ -24,62 +24,482 @@
+			if(("[c.overlay_refs[i]]" == "base") || ("[c.overlay_refs[i]]" == "second")) //setting up base layer overlay
+				var/overlay_layer
+				if("[c.overlay_refs[i]]" == "base")
+					if(src.clayer == 2)
+						overlay_layer = "second"
+						src.amount2 = c.amount
+					else if(src.clayer == 3)
+						overlay_layer = "third"
+						src.amount3 = c.amount
+				else if("[c.overlay_refs[i]]" == "second")
+					overlay_layer = "third"
+					src.amount3 = c.amount
+					src.clayer++
+					src.reagents.maximum_volume += 100
+					src.amount += 10
+				var/image/stack = new /image('icons/obj/foodNdrink/food_dessert.dmi',"cake[src.clayer]-overlay")
+				stack.color = (c.GetOverlayImage(c.overlay_refs[i])).color
+				src.UpdateOverlays(stack,"[overlay_layer]")
+				continue
+			var/image/buffer = c.GetOverlayImage("[c.overlay_refs[i]]") //generating the topping reference from the original cake to be stacked
+			var/list/tag
+			if(src.clayer == 2)
+				tag = replacetext("[c.overlay_refs[i]]","1","2")
+			else if(src.clayer == 3 && singlecake)
+				tag = replacetext("[c.overlay_refs[i]]","1","3")
+			else
+				tag = replacetext("[c.overlay_refs[i]]","2","3")
+			var/image/newoverlay = new /image('icons/obj/foodNdrink/food_dessert.dmi',"[tag]")
+			if(buffer.color)
+				newoverlay.color = buffer.color
+			src.UpdateOverlays(newoverlay,"[tag]")
+		if(c.litfam)
+			src.ignite()
+		qdel(c)
+
+
+	proc/build_cake(var/obj/item/cake_transfer,var/mob/user,var/mode,var/layer_tag,var/replacetext)//cake_transfer : passes a reference to the cake that we are building //mode : 1 or 2 : cake or slice //layer_tag and replacetext : references used with slicing //decompiles a full cake into slices or other cakes
+		if((mode<1) || (mode>2))
+			return
+		var/staticiterator = src.overlays.len
+		var/toggleswitch
+		var/slices
+		var/candle //cute little wax stick that people light on fire for their own enjoyment <3
+		var/obj/item/reagent_containers/food/snacks/cake/custom/cake
+		if(istype(cake_transfer,/obj/item/reagent_containers/food/snacks/cake/custom))
+			cake = cake_transfer
+		for(var/i=1,i<=staticiterator,i++)
+			if(mode == 1)
+				if("[src.overlay_refs[i]]" == "base")
+					continue
+				if(("[src.overlay_refs[i]]" == "second") || ("[src.overlay_refs[i]]" == "third"))
+					if(("[src.overlay_refs[i]]" == "second") && (src.clayer == 2))
+						toggleswitch = 1
+						cake.amount = src.amount2
+						src.amount2 = 0
+					else if(("[src.overlay_refs[i]]" == "third") && (src.clayer == 3))
+						toggleswitch = 1
+						cake.amount = src.amount3
+						src.amount3 = 0
+					if(toggleswitch)
+						var/image/stack = new /image('icons/obj/foodNdrink/food_dessert.dmi',"cake1-overlay")
+						stack.color = (src.GetOverlayImage(src.overlay_refs[i])).color
+						cake.UpdateOverlays(stack,"base")
+						src.ClearSpecificOverlays("[src.overlay_refs[i]]")
+						staticiterator--
+						i--
+						continue
```

Comment:
Same stuff as above applies (for in loop instead of the current thing is better, you need to put the result of GetOverlayImage into a `var/image` var first before getting `.color`).

Reply:
Actually the code seems so similar that maybe both of those parts of the code could instead be a separate proc to avoid code duplication? But also maybe not, I haven't looked into exact differences.

Reply:
Eh, probably scratch that, it seems sufficiently different further below.